test/cpp/api/support.h

```diff
@@ -49,6 +50,83 @@ inline bool pointer_equal(at::Tensor first, at::Tensor second) {
   return first.data_ptr<float>() == second.data_ptr<float>();
 }
 
+// This mirrors the `isinstance(x, torch.Tensor) and isinstance(y, Number)` branch
+// in `TestCase.assertEqual` in test/common_utils.py
+#define TENSOR(T, S) \
+inline void assert_tensor_equal(at::Tensor a, T b) { \
+  ASSERT_TRUE(std::abs(a.item<T>() - b) < 1e-4); \
+}
+AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
+#undef TENSOR
+
+// This mirrors the `isinstance(x, torch.Tensor) and isinstance(y, torch.Tensor)` branch
+// in `TestCase.assertEqual` in test/common_utils.py
+inline void assert_tensor_equal(at::Tensor a, at::Tensor b, bool allow_inf=false) {
+  ASSERT_TRUE(a.sizes() == b.sizes());
+  if (a.numel() > 0) {
+    if (a.device().type() == torch::kCPU && (a.dtype() == torch::kFloat16 || a.dtype() == torch::kBFloat16)) {
+      // CPU half and bfloat16 tensors don't have the methods we need below
+      a = a.to(torch::kFloat32);
+    }
+    b = b.to(a);
+
+    if ((a.dtype() == torch::kBool) != (b.dtype() == torch::kBool)) {
```

Comment:
I feel that it might be a good idea to keep it in sync with the Python version to make maintenance easier https://github.com/pytorch/pytorch/blob/638e4ad8b95793876c0f5e82db58c18618fb6e61/test/common_utils.py#L804 If the Python version changes, we can change the C++ version correspondingly :D